File: 99-Continue/158/CountalBinaryTree.py
from collections import deque

class Node:
    def __init__(self,data:int):
        self.left = self.right = None
        self.data = data

class BinaryTree:
    def __init__(self):
        self.root = None;
        
    def buildTree(self,arr:list[int]) -> Node:
        if len(arr)==0:
            return self.root
        
        self.root = Node(arr[0])
        queue = deque()
        queue.append(self.root)
        
        i = 1
        while queue and i<len(arr):
            current = queue.popleft()
            if i<len(arr):
                current.left = Node(arr[i])
                queue.append(current.left)
                i+=1
            if i<len(arr):
                current.right = Node(arr[i])
                queue.append(current.right)
                i+=1
        return self.root
    
    def countNodes(self) -> int:
        if self.root is None:
            return 0;
        
        count =1
        queue = deque([self.root])
        
        while queue:
            current = queue.popleft()
            if current.left is not None:
                queue.append(current.left)
                count+=1
            if current.right is not None:
                queue.append(current.right)
                count+=1
        return count
        

def main():
    n = int(input())
    lst = list(map(int,input().split()))
    tree = BinaryTree()
    tree.buildTree(lst)
    print(tree.countNodes())
    
if __name__=="__main__":
    main()

File: 99-Continue/158/CountalBinaryTree.java
import java.util.*;

public class CountalBinaryTree {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        BinaryTree tree  = new BinaryTree();
        tree.buildTree(n, sc);
        System.out.println(tree.countNodes());
        sc.close();
    }
}

class Node{
    Node left, right;
    int data;
    Node(int val){
        this.data = val;
    }
}

class BinaryTree{
    Node root;
    BinaryTree(){
        this.root = null;
    }
    
    Node buildTree(int n, Scanner sc){
        if(n==0){
            return root;
        }
        root = new Node(sc.nextInt());
        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);
        int i =1;
        while (!queue.isEmpty() && i<n){
            Node current = queue.poll();
            
            if(i<n){
                current.left = new Node(sc.nextInt());
                queue.offer(current.left);
                i++;
            }
            if(i<n){
                current.right = new Node(sc.nextInt());
                queue.offer(current.right);
                i++;
            }
        }
   
        return root;
    }
    
    int countNodes(){
        int count=0;
        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);
        count++;
        while(!queue.isEmpty()){
            Node current = queue.poll(); 
            if(current.left!=null){
                queue.offer(current.left);
                count++;
            }
            if(current.right!=null){
                queue.offer(current.right);
                count++;
            }
        }
        return count;
    }
}

